### PR TITLE
Support VDPA alongside SR-IOV on OVN NICs

### PR DESCRIPTION
**Pull request by gabrielmougard**
- [x] Add VDPA device tracking to resources API containing additional fields for `name` (`vdpa<ID>`) and a `device` (`vhost-vdpa-<ID>`, in /sys/bus/vdpa/drivers/vhost_vdpa/vdpa*)
- [x] Grow the `acceleration=="vdpa"` logic in `ovn_nic.go`
- [x] Utilities to create vDPA mgmt devices (need to be tested now)
- [x] Pass parameters to QEMU
- [x] Handle postStop in QEMU to setup the volatile["last_state.vdpa.name"]
- [x] Populate `ResourcesNetworkCardVDPA` in resources API
- [ ] Investigate QMP interface to switch an existing network card between vhost-vdpa and vhost-user at run time

### CODE REVIEW

**Comment by gabrielmougard** [comment]
@tomponline I just wanted to show here how we could avoid using `iproute2` (therefore no need to ship it in the snap) but rather talk to netlink to create the mgmt vDPA device.

**Comment by tomponline** [comment]
@gabrielmougard We've seen attempts to use that netlink package in the past for other PRs.
Your PR doesn't include the go mod update required yet, so won't build, so its not possible to see what other dependencies that pulls in.

That being said, I think it would be preferable to use netlink where possible to, as you say, reduce dependencies on external programs (and also offers the long-discussed possibility of reducing/removing our dependency on the `ip` command also). 

What do you think about this approach @stgraber ?

**Comment by gabrielmougard** [comment]
@tomponline the dep is already there

**Comment by tomponline** [comment]
> @tomponline the dep is already there

Well that is convenient, its pulled in as an indirect then by some other dependency?

**Comment by gabrielmougard** [comment]
> > @tomponline the dep is already there
> 
> Well that is convenient! Its pulled in as an indirect then by some other dependency?

yep

**Comment by gabrielmougard** [comment]
@stgraber the `ip/vdpa.go` lib is working well :)

**Comment by gabrielmougard** [comment]
@stgraber @tomponline Oooh the test server has probably an old version of iproute2 that's why

**Comment by gabrielmougard** [comment]
I can't make the following library fully work (can't pass `maxVQP` and `MTU` to `netlink` for an unknown reason. Or rather, I pass these packets in the netlink header, but the kernel ignores them. I had a hard time debugging with `strace`, so I think I'll postpone this for now). I paste it there for now (so that we have a trace of it for an other PR) and I switch back to using the `iproute2` utility tool `vdpa`.

```go
package ip

import (
	"encoding/binary"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"syscall"

	"github.com/vishvananda/netlink"
	"github.com/vishvananda/netlink/nl"
)

// vDPA Netlink name.
const (
	VDPAGenlName = "vdpa"
)

// vDPA Netlink command.
const (
	VDPACmdUnspec uint8 = iota
	VDPACmdMgmtDevNew
	VDPACmdMgmtDevGet
	VDPACmdDevNew
	VDPACmdDevDel
	VDPACmdDevGet
	VDPACmdDevConfigGet
)

// VDPA Netlink Attributes.
const (
	VDPAAttrUnspec = iota

	// bus name (optional) + dev name together make the parent device handle.
	VDPAAttrMgmtDevBusName          // string
	VDPAAttrMgmtDevDevName          // string
	VDPAAttrMgmtDevSupportedClasses // u64

	VDPAAttrDevName      // string
	VDPAAttrDevID        // u32
	VDPAAttrDevVendorID  // u32
	VDPAAttrDevMaxVqs    // u32
	VDPAAttrDevMaxVqSize // u16
	VDPAAttrDevMinVqSize // u16

	VDPAAttrDevNetCfgMacAddr // binary
	VDPAAttrDevNetStatus     // u8
	VDPAAttrDevNetCfgMaxVqp  // u16
	VDPAAttrGetNetCfgMTU     // u16

	VDPAAttrDevNegociatedFeatures    /* u64 */
        VDPAAttrDevMgmtDevMaxVqs        /* u32 */
        /* virtio features that are supported by the vDPA management device */
        VDPAAttrDevSupportedFeatures    /* u64 */

        VDPAAttrDevQueueIndex              /* u32 */
        VDPAAttrDevVendorAttrName        /* string */
        VDPAAttrDevVendorAttrValue        /* u64 */

        /* virtio features that are provisioned to the vDPA device */
        VDPAAttrDevFeatures	                 /* u64 */

	// new attributes must be added above here.
	VDPAAttrMax
)

// Base flags passed to all Netlink requests.
const (
	commonNLFlags = syscall.NLM_F_REQUEST | syscall.NLM_F_ACK
)

// MAX_VQP is the maximum number of VQPs supported by the vDPA device and is always the same as of now.
const (
	vDPAMaxVQP = uint16(16)
)

// VDPA device classes.
const (
	vdpaBusDevDir   = "/sys/bus/vdpa/devices"
	vdpaVhostDevDir = "/dev"
)

// VhostVdpa is the vhost-vdpa device information.
type VhostVDPA struct {
	Name string
	Path string
}

// MgmtVDPADev represents the vDPA management device information.
type MgmtVDPADev struct {
	BusName string // e.g. "pci"
	DevName string // e.g. "0000:00:08.2"
}

// VDPADev represents the vDPA device information.
type VDPADev struct {
	// Name of the vDPA created device. e.g. "vdpa0" (note: the iproute2 associated command would look like `vdpa dev add mgmtdev pci/<PCI_SLOT_NAME> name vdpa0 max_vqp <MAX_VQP>`).
	Name string
	// Max VQP supported by the vDPA device.
	MaxVQP uint16
	// MTU of the vDPA device.
	MTU uint16
	// Associated vDPA management device.
	MgmtDev *MgmtVDPADev
	// Associated vhost-vdpa device.
	VhostVDPA *VhostVDPA
}

// ParseAttributes parses the attributes of a netlink message for a vDPA management device.
func (d *MgmtVDPADev) parseAttributes(attrs []syscall.NetlinkRouteAttr) error {
	for _, attr := range attrs {
		switch attr.Attr.Type {
		case VDPAAttrMgmtDevBusName:
			d.BusName = string(attr.Value[:len(attr.Value)-1])
		case VDPAAttrMgmtDevDevName:
			d.DevName = string(attr.Value[:len(attr.Value)-1])
		}
	}
	return nil
}

// getVhostVDPADevInPath returns the VhostVDPA found in the provided parent device's path.
func getVhostVDPADevInPath(parentPath string) (*VhostVDPA, error) {
	fd, err := os.Open(parentPath)
	if err != nil {
		return nil, err
	}

	defer fd.Close()

	fileInfos, err := fd.Readdir(-1)
	if err != nil {
		return nil, err
	}

	for _, file := range fileInfos {
		if strings.Contains(file.Name(), "vhost-vdpa") && file.IsDir() {
			devicePath := filepath.Join(vdpaVhostDevDir, file.Name())
			info, err := os.Stat(devicePath)
			if err != nil {
				return nil, err
			}

			if info.Mode()&os.ModeDevice == 0 {
				return nil, fmt.Errorf("vhost device %s is not a valid device", devicePath)
			}

			return &VhostVDPA{
				Name: file.Name(),
				Path: devicePath,
			}, nil
		}
	}

	return nil, fmt.Errorf("no vhost-vdpa device found in %s", parentPath)
}

// ParseAttributes parses the attributes of a netlink message for a vDPA device.
func (d *VDPADev) parseAttributes(attrs []syscall.NetlinkRouteAttr) error {
	d.MgmtDev = &MgmtVDPADev{}
	for _, attr := range attrs {
		switch attr.Attr.Type {
		case VDPAAttrDevName:
			d.Name = string(attr.Value[:len(attr.Value)-1])
		case VDPAAttrDevNetCfgMaxVqp:
			d.MaxVQP = binary.BigEndian.Uint16(attr.Value[:len(attr.Value)-1])
		case VDPAAttrGetNetCfgMTU:
			d.MTU = binary.BigEndian.Uint16(attr.Value[:len(attr.Value)-1])
		case VDPAAttrMgmtDevBusName:
			d.MgmtDev.BusName = string(attr.Value[:len(attr.Value)-1])
		case VDPAAttrMgmtDevDevName:
			d.MgmtDev.DevName = string(attr.Value[:len(attr.Value)-1])
		}
	}

	// Get the vhost-vdpa device associated with the vDPA device.
	vhostVDPA, err := getVhostVDPADevInPath(filepath.Join(vdpaBusDevDir, d.Name))
	if err != nil {
		return err
	}

	d.VhostVDPA = vhostVDPA
	return nil
}

// runVDPANetlinkCmd executes a vDPA netlink command and returns the response.
func runVDPANetlinkCmd(command uint8, flags int, data []*nl.RtAttr) ([][]byte, error) {
	f, err := netlink.GenlFamilyGet(VDPAGenlName)
	if err != nil {
		return nil, err
	}

	msg := &nl.Genlmsg{
		Command: command,
		Version: nl.GENL_CTRL_VERSION,
	}

	req := nl.NewNetlinkRequest(int(f.ID), commonNLFlags|flags)
	// Pass the data into the request header
	req.AddData(msg)
	for _, d := range data {
		req.AddData(d)
	}

	// Execute the request
	msgs, err := req.Execute(syscall.NETLINK_GENERIC, 0)
	if err != nil {
		return nil, err
	}

	return msgs, nil
}

// newNetlinkAttribute creates a new netlink attribute based on the attribute type and data.
func newNetlinkAttribute(attrType int, data interface{}) (*nl.RtAttr, error) {
	switch attrType {
	case VDPAAttrMgmtDevBusName, VDPAAttrMgmtDevDevName, VDPAAttrDevName:
		strData, ok := data.(string)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires string data", attrType)
		}

		bytes := make([]byte, len(strData)+1)
		copy(bytes, strData)
		return nl.NewRtAttr(attrType, bytes), nil
	case VDPAAttrMgmtDevSupportedClasses:
		u64Data, ok := data.(uint64)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires uint64 data", attrType)
		}

		return nl.NewRtAttr(attrType, nl.Uint64Attr(u64Data)), nil
	case VDPAAttrDevID, VDPAAttrDevVendorID, VDPAAttrDevMaxVqs:
		u32Data, ok := data.(uint32)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires uint32 data", attrType)
		}

		return nl.NewRtAttr(attrType, nl.Uint32Attr(u32Data)), nil
	case VDPAAttrDevMaxVqSize, VDPAAttrDevMinVqSize, VDPAAttrDevNetCfgMaxVqp, VDPAAttrGetNetCfgMTU:
		u16Data, ok := data.(uint16)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires uint16 data", attrType)
		}

		return nl.NewRtAttr(attrType, nl.Uint16Attr(u16Data)), nil
	case VDPAAttrDevNetStatus:
		u8Data, ok := data.(uint8)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires uint8 data", attrType)
		}

		return nl.NewRtAttr(attrType, nl.Uint8Attr(u8Data)), nil
	case VDPAAttrDevNetCfgMacAddr:
		binData, ok := data.([]byte)
		if !ok {
			return nil, fmt.Errorf("Attribute type %d requires []byte data", attrType)
		}

		return nl.NewRtAttr(attrType, binData), nil
	default:
		return nil, fmt.Errorf("Unknown attribute type %d", attrType)
	}
}

// parseMgmtVDPADevList parses a list of vDPA management device netlink messages.
func parseMgmtVDPADevList(msgs [][]byte) ([]*MgmtVDPADev, error) {
	devices := make([]*MgmtVDPADev, 0, len(msgs))
	for _, m := range msgs {
		attrs, err := nl.ParseRouteAttr(m[nl.SizeofGenlmsg:])
		if err != nil {
			return nil, err
		}

		dev := &MgmtVDPADev{}
		if err = dev.parseAttributes(attrs); err != nil {
			return nil, err
		}

		devices = append(devices, dev)
	}

	return devices, nil
}

// parseVDPADevList parses a list of vDPA device netlink messages.
func parseVDPADevList(msgs [][]byte) ([]*VDPADev, error) {
	devices := make([]*VDPADev, 0, len(msgs))
	for _, m := range msgs {
		attrs, err := nl.ParseRouteAttr(m[nl.SizeofGenlmsg:])
		if err != nil {
			return nil, err
		}

		dev := &VDPADev{}
		if err = dev.parseAttributes(attrs); err != nil {
			return nil, err
		}

		devices = append(devices, dev)
	}

	return devices, nil
}

// ListVDPAMgmtDevices returns the list of all vDPA management devices.
func ListVDPAMgmtDevices() ([]*MgmtVDPADev, error) {
	resp, err := runVDPANetlinkCmd(VDPACmdMgmtDevGet, syscall.NLM_F_DUMP, nil)
	if err != nil {
		return nil, err
	}

	mgtmDevs, err := parseMgmtVDPADevList(resp)
	if err != nil {
		return nil, err
	}

	return mgtmDevs, nil
}

// ListVDPADevices returns the list of all vDPA devices.
func ListVDPADevices() ([]*VDPADev, error) {
	resp, err := runVDPANetlinkCmd(VDPACmdDevGet, syscall.NLM_F_DUMP, nil)
	if err != nil {
		return nil, err
	}

	devices, err := parseVDPADevList(resp)
	if err != nil {
		return nil, err
	}

	return devices, nil
}

// AddVDPADevice adds a new vDPA device.
func AddVDPADevice(pciDevSlotName string) (*VDPADev, error) {
	// List existing vDPA devices
	vdpaDevs, err := ListVDPADevices()
	if err != nil {
		return nil, err
	}

	existingVDPADevNames := make(map[string]struct{})
	for _, vdpaDev := range vdpaDevs {
		existingVDPADevNames[vdpaDev.Name] = struct{}{}
	}

	// Create the netlink attributes
	header := []*nl.RtAttr{}
	busName, err := newNetlinkAttribute(VDPAAttrMgmtDevBusName, "pci")
	if err != nil {
		return nil, err
	}

	header = append(header, busName)

	mgmtDevDevName, err := newNetlinkAttribute(VDPAAttrMgmtDevDevName, pciDevSlotName)
	if err != nil {
		return nil, err
	}

	header = append(header, mgmtDevDevName)

	// Generate a unique attribute name for the vDPA device (i.e, vdpa0, vdpa1, etc.)
	baseVDPAName, idx, generatedVDPADevName := "vdpa", 0, ""
	for {
		generatedVDPADevName = fmt.Sprintf("%s%d", baseVDPAName, idx)
		_, ok := existingVDPADevNames[generatedVDPADevName]
		if !ok {
			break
		}

		idx++
	}

	devName, err := newNetlinkAttribute(VDPAAttrDevName, generatedVDPADevName)
	if err != nil {
		return nil, err
	}

	header = append(header, devName)

	maxVQP, err := newNetlinkAttribute(VDPAAttrDevNetCfgMaxVqp, vDPAMaxVQP)
	if err != nil {
		return nil, err
	}

	header = append(header, maxVQP)

	_, err = runVDPANetlinkCmd(VDPACmdDevNew, 0, header)
	if err != nil {
		return nil, err
	}

	// Now that the vDPA device has been created in the kernel, return the VDPADev struct
	msgs, err := runVDPANetlinkCmd(VDPACmdDevGet, 0, []*nl.RtAttr{devName})
	if err != nil {
		return nil, err
	}

	vdpaDevs, err = parseVDPADevList(msgs)
	if err != nil {
		return nil, err
	}

	return vdpaDevs[0], nil
}

// DeleteVDPADevice deletes a vDPA management device.
func DeleteVDPADevice(vDPADevName string) error {
	header := []*nl.RtAttr{}
	devName, err := newNetlinkAttribute(VDPAAttrDevName, vDPADevName)
	if err != nil {
		return err
	}

	header = append(header, devName)

	_, err = runVDPANetlinkCmd(VDPACmdDevDel, 0, header)
	if err != nil {
		return err
	}

	return nil
}
```

File: lxd/ip/vdpa.go [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+)
+
+// vDPA Netlink name.
+const (
+	VDPAGenlName = "vdpa"
+)
+
+// vDPA Netlink command.
+const (
+	VDPACmdUnspec uint8 = iota
```
Comment (tomponline) [inline lxd/ip/vdpa.go:22]: @gabrielmougard another "general" code style comment here (I'll wait until you mark this ready for review before doing something more thorough). But generally I would suggest you unexport any variables/constants/functions unless they are actually needed to be used outside of the package. This helps to communicate to the reviewer what the external API is of your code. Thanks

**Comment by tomponline** [comment]
Looks like some very minor static analysis failures.

File: lxd/ip/vdpa.go (hunk 1 above)
Comment (tomponline) [reply lxd/ip/vdpa.go:22]: Looks like a lot of these constants are still unnecessarily exported? We shouldn't be exporting netlink concepts outside of the package.

File: shared/api/resource.go [hunk 2]
```diff
@@ -551,6 +554,19 @@ type ResourcesNetworkCardSRIOV struct {
 	VFs []ResourcesNetworkCard `json:"vfs" yaml:"vfs"`
 }
 
+// ResourceNetworkCardVDPA represents the VDPA configuration of the network card
+//
+// swagger:model
+//
+// API extension: resources_v2.
```
Comment (tomponline) [inline shared/api/resource.go:561]: Is this API extension right? I think it should be its own separate API extension.

File: lxd/device/nic_ovn.go [hunk 3]
```diff
@@ -441,6 +442,65 @@ func (d *nicOVN) Start() (*deviceConfig.RunConfig, error) {
...
+			vDPADevice, err = ip.AddVDPADevice(vfPCIDev.SlotName, saveData)
+			if err != nil {
+				network.SRIOVVirtualFunctionMutex.Unlock()
+				return nil, err
+			}
+
+			network.SRIOVVirtualFunctionMutex.Unlock()
+
+			// Setup the guest network interface.
+			if d.inst.Type() == instancetype.Container {
```
Comment (tomponline) [inline lxd/device/nic_ovn.go:497]: @gabrielmougard @stgraber `acceleration=vdpa` cannot be used with containers right? So should be failing earlier during 2nd stage validation (when d.inst != nil) if the instance is a container. Or should we silently fallback to doing `acceleration=sriov` mode like it looks like its doing here.

**Comment by tomponline** [comment]
@gabrielmougard it doesn't look like you've handle removing the representor port from the integration bridge like here:

https://github.com/lxc/lxd/blob/master/lxd/device/nic_ovn.go#L699-L705

File: lxd/device/nic_ovn.go [hunk 4]
```diff
@@ -747,6 +822,7 @@ func (d *nicOVN) postStop() error {
 			"last_state.hwaddr":        "",
 			"last_state.mtu":           "",
 			"last_state.created":       "",
+			"last_state.vdpa.name":     "",
```
Comment (tomponline) [inline lxd/device/nic_ovn.go:825]: We need this added to the docs in doc/reference/instance_options.md, search for "last_state.vf.id" for an example.

**Comment by tomponline** [comment]
> @gabrielmougard it doesn't look like you've handle removing the representor port from the integration bridge like here:
> 
> https://github.com/lxc/lxd/blob/master/lxd/device/nic_ovn.go#L699-L705

Oh no I see it. Thanks

File: lxd/device/nic_ovn.go (hunk 4 above)
Comment (gabrielmougard) [reply lxd/device/nic_ovn.go:825]: sure

**Comment by tomponline** [comment]
@gabrielmougard please can you run `make update-api` and add that as a separate commit.

File: shared/api/resource.go (hunk 2 above)
Comment (tomponline) [reply shared/api/resource.go:561]: Should be `ovn_nic_acceleration_vdpa` I think.

File: doc/rest-api.yaml [hunk 5]
```diff
@@ -4520,6 +4522,19 @@ definitions:
                 x-go-name: VFs
         type: object
         x-go-package: github.com/lxc/lxd/shared/api
+    ResourcesNetworkCardVDPA:
```
Comment (tomponline) [inline doc/rest-api.yaml:4525]: We normally do this in a separate commit.

File: lxd/ip/vdpa.go (hunk 1 above)
Comment (gabrielmougard) [reply lxd/ip/vdpa.go:22]: As I use `iota` and that the order matter (these are kernel bit flags on the netlink side), it might cause an issue. If the parameter is at the end of the `iota` block and is unused then it's ok, else it's not.. Or I just don't use `iota` to avoid all this.. @tomponline what do you think ?

Comment (tomponline) [reply lxd/ip/vdpa.go:22]: I dont see how that is relevant to them being exported or not though?

Comment (tomponline) [reply lxd/ip/vdpa.go:22]: By exported I mean https://go.dev/tour/basics/3

File: lxd/apparmor/instance_qemu.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
   /dev/net/tun                              rw,
   /dev/ptmx                                 rw,
   /dev/sev                                  rw,
+  /dev/vhost-vdpa-*                         rw,
```
Comment (tomponline) [inline lxd/apparmor/instance_qemu.go:27]: This isn't needed now that @stgraber has added `vhostfd` support.

File: lxd/instance/drivers/driver_qemu.go [hunk 7]
```diff
@@ -3853,6 +3857,41 @@ func (d *qemu) addNetDevConfig(busName string, qemuDev map[string]string, bootIn
...
+
+			err := os.Chown(vhostVDPAPath, int(d.state.OS.UnprivUID), -1)
+			if err != nil {
+				return fmt.Errorf("Failed to chown vDPA device %q: %w", vhostVDPAPath, err)
+			}
+
+			qemuNetDev := map[string]any{
+				"id":       fmt.Sprintf("vhost-%s", vDPADevName),
+				"type":     "vhost-vdpa",
+				"vhostdev": vhostVDPAPath,
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:3875]: We need to change this to use `vhostfd` like you had it earlier (i.e using `monitor.SendFile()` like `macvlan`).

File: shared/version/api.go [hunk 8]
```diff
@@ -368,6 +368,7 @@ var APIExtensions = []string{
 	"migration_vm_live",
 	"ovn_nic_nesting",
 	"oidc",
+	"ovn_nic_acceleration_vdpa",
```
Comment (tomponline) [inline shared/version/api.go:371]: This commit should be include the changes to [doc/api-extensions.md](https://github.com/lxc/lxd/pull/11528/commits/556eda5c2e22150880aa8e063bb0611ba4527a66#diff-906870407147ce5fd1c4850626a492df3cd564c6981706e8ef8f7247e9200637) and the commit message should be:

`api: Add the ovn_nic_acceleration_vdpa extension`

File: doc/api-extensions.md [hunk 9]
```diff
@@ -2201,3 +2201,7 @@ This adds the following new configuration keys:
 * `oidc.issuer`
 * `oidc.client.id`
 * `oidc.audience`
+
+## `ovn_nic_acceleration_vdpa`
```
Comment (tomponline) [inline doc/api-extensions.md:2205]: Lets move this bit into the api extension commit.

File: doc/api-extensions.md [hunk 10]
```diff
@@ -2201,3 +2201,7 @@ This adds the following new configuration keys:
...
+
+This updates the `ovn_nic_acceleration` API extension. The `acceleration` configuration key for OVN NICs can now takes the value `vdpa` to support Virtual Data Path Acceleration (VDPA).
```
Comment (tomponline) [inline doc/api-extensions.md:2207]: I suspect "OVN" here will need to be `ovn` to appease the spellcheck.

File: lxd/instance/drivers/driver_qemu.go (hunk 7 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:3875]: yep testing it right now

File: shared/api/resource.go [hunk 11]
```diff
@@ -551,6 +554,19 @@ type ResourcesNetworkCardSRIOV struct {
 	VFs []ResourcesNetworkCard `json:"vfs" yaml:"vfs"`
 }
 
+// ResourceNetworkCardVDPA represents the VDPA configuration of the network card
+//
+// swagger:model
+//
+// API extension: ovn_nic_acceleration_vdpa.
+type ResourcesNetworkCardVDPA struct {
```
Comment (tomponline) [inline shared/api/resource.go:562]: @gabrielmougard unless I'm mistaken, this PR doesn't yet populate this struct?

File: lxd/device/nic_ovn.go (hunk 3 above)
Comment (tomponline) [reply lxd/device/nic_ovn.go:497]: @stgraber has confirmed:

> I think we should prevent it and then add an "auto" mode later on

**Comment by tomponline** [comment]
@gabrielmougard multiple static analysis failures:

```
Error: lxd/ip/vdpa.go:22:2: `vDPACmdUnspec` is unused (deadcode)
	vDPACmdUnspec uint8 = iota
	^
Error: lxd/ip/vdpa.go:23:2: `vDPACmdMgmtDevNew` is unused (deadcode)
	vDPACmdMgmtDevNew
	^
Error: lxd/ip/vdpa.go:28:2: `vDPACmdDevConfigGet` is unused (deadcode)
	vDPACmdDevConfigGet
	^
Error: lxd/ip/vdpa.go:33:2: `vDPAAttrUnspec` is unused (deadcode)
	vDPAAttrUnspec = iota
	^
make: *** [Makefile:267: static-analysis] Error 1
```

File: shared/api/resource.go (hunk 11 above)
Comment (gabrielmougard) [reply shared/api/resource.go:562]: no indeed

Comment (gabrielmougard) [reply shared/api/resource.go:562]: I'll fix that
